### PR TITLE
[SysCall] add support to enum methods,

### PR DESCRIPTION
**Pull request by Jim8y**
This pr adds a group of enum methods.

### CODE REVIEW

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs [hunk 1]
```diff
@@ -730,6 +742,20 @@ private static void RegisterNullableTypeHandlers()
...
+        // RegisterHandler((ushort? x, ushort defaultValue) => x.GetValueOrDefault(defaultValue), HandleNullableUShortGetValueOrDefaultWithParam);
+        // RegisterHandler((int? x, int defaultValue) => x.GetValueOrDefault(defaultValue), HandleNullableIntGetValueOrDefaultWithParam);
+        // RegisterHandler((uint? x, uint defaultValue) => x.GetValueOrDefault(defaultValue), HandleNullableUIntGetValueOrDefaultWithParam);
+        // RegisterHandler((long? x, long defaultValue) => x.GetValueOrDefault(defaultValue), HandleNullableLongGetValueOrDefaultWithParam);
+        // RegisterHandler((ulong? x, ulong defaultValue) => x.GetValueOrDefault(defaultValue), HandleNullableULongGetValueOrDefaultWithParam);
+        // RegisterHandler((BigInteger? x, BigInteger defaultValue) => x.GetValueOrDefault(defaultValue), HandleNullableBigIntegerGetValueOrDefaultWithParam);
+        // RegisterHandler((char? x, char defaultValue) => x.GetValueOrDefault(defaultValue), HandleNullableCharGetValueOrDefaultWithParam);
+        // RegisterHandler((bool? x, bool defaultValue) => x.GetValueOrDefault(defaultValue), HandleNullableBoolGetValueOrDefaultWithParam);
+        #endregion GetValueOrDefault
```
Comment (Jim8y) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs:758]:
```suggestion
```

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs [hunk 2]
```diff
@@ -618,12 +628,14 @@ private static void RegisterCharHandlers()
...
 
-        // Add missing char methods
         RegisterHandler((char c) => char.ToLowerInvariant(c), HandleCharToLowerInvariant);
         RegisterHandler((char c) => char.ToUpperInvariant(c), HandleCharToUpperInvariant);
         RegisterHandler((char c) => char.IsAscii(c), HandleCharIsAscii);
         RegisterHandler((char c) => char.IsAsciiDigit(c), HandleCharIsAsciiDigit);
         RegisterHandler((char c) => char.IsAsciiLetter(c), HandleCharIsAsciiLetter);
+        RegisterHandler((string s) => char.Parse(s), HandleCharParse);
+        // RegisterHandler((string s, char result) => char.TryParse(s, out result), HandleCharTryParse);
+
```
Comment (Jim8y) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs:638]:
```suggestion
```

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs [hunk 3]
```diff
@@ -584,6 +587,13 @@ private static void RegisterStringHandlers()
         RegisterHandler((string s, char trimChar) => s.Trim(trimChar), HandleStringTrimChar);
         RegisterHandler((string s, string oldValue, string newValue) => s.Replace(oldValue, newValue), HandleStringReplace);
         RegisterHandler((string s) => s.Length, HandleStringLength);
+
+        // RegisterHandler((string s, char[] separator, StringSplitOptions options) => s.Split(separator, options), HandleStringSplit);
+        // RegisterHandler((string separator, IEnumerable<string?> values) => string.Join(separator, values), HandleStringJoin);
+        // RegisterHandler((string s, string value) => s.StartsWith(value), HandleStringStartsWith);
+        // RegisterHandler((string s, int totalWidth, char paddingChar) => s.PadLeft(totalWidth, paddingChar), HandleStringPadLeft);
+        // RegisterHandler((string s, int totalWidth, char paddingChar) => s.PadRight(totalWidth, paddingChar), HandleStringPadRight);
+        // RegisterHandler((string s, char[] trimChars) => s.Trim(trimChars), HandleStringTrimWithCharArray);
```
Comment (Jim8y) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs:596]:
```suggestion
```

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.cs [hunk 4]
```diff
@@ -263,6 +272,8 @@ private bool TryProcessSystemMethods(SemanticModel model, IMethodSymbol symbol,
         var key = symbol.ToString()!.Replace("out ", "");
         key = (from parameter in symbol.Parameters let parameterType = parameter.Type.ToString() where !parameter.Type.IsValueType && parameterType!.EndsWith('?') select parameterType).Aggregate(key, (current, parameterType) => current.Replace(parameterType, parameterType[..^1]));
         if (key == "string.ToString()") key = "object.ToString()";
+        if (key.Contains("System.Enum.GetName<")) key = "System.Enum.GetName<>()";
```
Comment (Jim8y) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.cs:275]: Can not directly process this, cause the name of the method is not the same as its signature that we invokes, so i have to add direct string key.

File: src/Neo.Compiler.CSharp/Diagnostic/DiagnosticId.cs [hunk 5]
```diff
@@ -36,5 +36,8 @@ static class DiagnosticId
         public const string InvalidInitialValue = "NC3005";
         public const string IncorrectNEPStandard = "NC3006";
         public const string CapturedStaticFieldNotFound = "NC3007";
+
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/Diagnostic/DiagnosticId.cs:39]:
```suggestion
```

File: tests/Neo.Compiler.CSharp.TestContracts/Contract_Enum.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
+using Neo.SmartContract.Framework.Attributes;
+using System.Collections;
+using System.Numerics;
+
+namespace Neo.Compiler.CSharp.TestContracts
+{
+
```
Comment (shargon) [inline tests/Neo.Compiler.CSharp.TestContracts/Contract_Enum.cs:7]:
```suggestion
```

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.cs [hunk 7]
```diff
@@ -60,10 +60,17 @@ private static void RegisterHandler<T1, T2, T3, TResult>(Expression<Func<T1, T2,
...
+        SystemCallHandlers[key] = handler;
+    }
+
+    private static void RegisterHandler<T1, T2, T3, T4, T5, TResult>(Expression<Func<T1, T2, T3, T4, T5, TResult>> expression, SystemCallHandler handler)
+    {
+        var key = GetKeyFromExpression(expression, typeof(T1), typeof(T2), typeof(T3), typeof(T4), typeof(T5));
         SystemCallHandlers[key] = handler;
     }
 
+
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.cs:73]:
```suggestion
```

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs [hunk 8]
```diff
@@ -747,4 +749,26 @@ private static void RegisterBitOperationsHandlers()
...
+        RegisterHandler((Type enumType, object value) => Enum.GetName(enumType, value), HandleEnumGetNameWithType, "System.Enum.GetName()");
+
+        // these two methods will not be supported, since we cannot apply format logic.
+        // RegisterHandler((Enum value) => Enum.Format(value.GetType(), value, "G"), HandleEnumFormat);
+        // RegisterHandler((Type enumType, object value, string format) => Enum.Format(enumType, value, format), HandleEnumFormatWithType);
+
+        // these two methods will not be supported, since we don't have `Type` class support in neo csharp.
+        // RegisterHandler((Enum value) => Enum.GetUnderlyingType(value.GetType()), HandleEnumGetUnderlyingType);
+        // RegisterHandler((Type enumType) => Enum.GetUnderlyingType(enumType), HandleEnumGetUnderlyingTypeWithType);
+
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs:772]:
```suggestion
```

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs [hunk 9]
```diff
@@ -747,4 +749,25 @@ private static void RegisterBitOperationsHandlers()
...
+        RegisterHandler((Type enumType, string value, object result) => Enum.TryParse(enumType, value, out result), HandleEnumTryParse);
+        RegisterHandler((Type enumType, string value, bool ignoreCase, object result) => Enum.TryParse(enumType, value, ignoreCase, out result), HandleEnumTryParseIgnoreCase);
+        RegisterHandler((Type enumType) => Enum.GetNames(enumType), HandleEnumGetNames);
+        RegisterHandler((Type enumType) => Enum.GetValues(enumType), HandleEnumGetValues);
+        RegisterHandler((Type enumType, object value) => Enum.IsDefined(enumType, value), HandleEnumIsDefined);
+        RegisterHandler((Type enumType, string name) => Enum.IsDefined(enumType, name), HandleEnumIsDefinedByName);
+        RegisterHandler((Enum value) => Enum.GetName(value.GetType(), value), HandleEnumGetName, "System.Enum.GetName<>()");
+        RegisterHandler((Type enumType, object value) => Enum.GetName(enumType, value), HandleEnumGetNameWithType, "System.Enum.GetName()");
+
+        // these two methods will not be supported, since we cannot apply format logic.
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs:765]: It's possible to detect them with the analyzer?

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.cs [hunk 10]
```diff
@@ -263,6 +271,8 @@ private bool TryProcessSystemMethods(SemanticModel model, IMethodSymbol symbol,
         var key = symbol.ToString()!.Replace("out ", "");
         key = (from parameter in symbol.Parameters let parameterType = parameter.Type.ToString() where !parameter.Type.IsValueType && parameterType!.EndsWith('?') select parameterType).Aggregate(key, (current, parameterType) => current.Replace(parameterType, parameterType[..^1]));
         if (key == "string.ToString()") key = "object.ToString()";
+        if (key.Contains("System.Enum.GetName<")) key = "System.Enum.GetName<>()";
+        if (key.Contains("System.Enum.GetName(")) key = "System.Enum.GetName()";
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.cs:275]: Change these two Contains for Startwith?

Comment (Jim8y) [reply src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.cs]: done

File: src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs (hunk 9 above)
Comment (Jim8y) [reply src/Neo.Compiler.CSharp/MethodConvert/System/SystemCall.Register.cs:765]: added

**Review by shargon — APPROVED** [review]

**Pull request merged**